### PR TITLE
Improve parquet `MetadataFetch` and `AsyncFileReader` docs

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Part of https://github.com/apache/arrow-rs/issues/6504

# Rationale for this change
 
While working on the example for using parquet metadata in a local cache (https://github.com/apache/arrow-rs/pull/6081) some of these traits and how to implement them were not clear to me

# What changes are included in this PR?

Add additional documentation and explanations

# Are there any user-facing changes?
Better documentation and examples. No functional changes

### CODE REVIEW

File: parquet/src/arrow/async_reader/metadata.rs [hunk 1]
```diff
@@ -28,8 +28,44 @@ use std::future::Future;
...
+///             .boxed() // turn into BoxedFuture, using FutureExt::boxed
+///     }
+/// }
+///```
 pub trait MetadataFetch {
-    /// Fetches a range of bytes asynchronously
+    /// Return a future that fetches the specified range of bytes asynchronously
+    ///
+    /// Note the returned type is a boxed future, often created by
+    /// [FutureExt::boxed]. See the trait documentation for an example
```
Comment (alamb) [inline parquet/src/arrow/async_reader/metadata.rs:68]: If you are not familiar with rust futures `BoxFuture` is somewhat exotic. I will also make a PR upstream in futures to try and improve the docs there

Comment (alamb) [reply parquet/src/arrow/async_reader/metadata.rs:68]: https://github.com/rust-lang/futures-rs/pull/2887 -- we'll see what the maintainers say

File: parquet/src/arrow/async_reader/metadata.rs [hunk 2]
```diff
@@ -28,8 +28,44 @@ use std::future::Future;
 use std::ops::Range;
 
 /// A data source that can be used with [`MetadataLoader`] to load [`ParquetMetaData`]
+///
+/// Note that implementation is are provided for [`AsyncFileReader`].
```
Comment (mbrobbel) [inline parquet/src/arrow/async_reader/metadata.rs:32]:
```suggestion
/// Note that implementation is provided for [`AsyncFileReader`].
```

File: parquet/src/arrow/async_reader/mod.rs [hunk 3]
```diff
@@ -121,6 +121,18 @@ use crate::arrow::schema::ParquetField;
...
 
 /// The asynchronous interface used by [`ParquetRecordBatchStream`] to read parquet files
+///
+/// Notes:
+///
+/// 1. There is a default implementation for types that implement [`AsyncRead`]
+///    and [`AsyncSeek`], for example [`tokio::fs::File`].
+///
+/// 2. [`ParquetObjectReader`], available when the `object_store` crate feature
+///    is activated, implements this interface for [`ObjectStore`].
```
Comment (mbrobbel) [inline parquet/src/arrow/async_reader/mod.rs:131]:
```suggestion
///    is enabled, implements this interface for [`ObjectStore`].
```

**Review by etseidl — APPROVED** [review]
Looks good to me. 👍  @mbrobbel found the only issues I did.

File: parquet/src/arrow/async_reader/metadata.rs [hunk 4]
```diff
@@ -28,8 +28,44 @@ use std::future::Future;
...
+/// // Adapter that implements the API for reading bytes from an async source (in
+/// // this case a tokio::fs::File)
+/// struct TokioFileMetadata {
+///     file: tokio::fs::File,
+/// }
+/// impl MetadataFetch for TokioFileMetadata {
+///     fn fetch(&mut self, range: Range<usize>) -> BoxFuture<'_, Result<Bytes>> {
+///         // return a future that fetches data in range
+///         async move {
+///             let mut buf = vec![0; range.end - range.start]; // target buffer
```
Comment (progval) [inline parquet/src/arrow/async_reader/metadata.rs:54]:
```suggestion
///             let mut buf = vec![0; range.len()]; // target buffer
```

File: parquet/src/file/metadata/reader.rs [hunk 5]
```diff
@@ -39,11 +39,18 @@ use crate::arrow::async_reader::MetadataFetch;
 /// See [`crate::file::metadata::ParquetMetaDataWriter#output-format`] for a description of
 /// the Parquet metadata.
 ///
+/// Parquet metadata is not necessarily contiguous in the files: part is stored
+/// in the footer (the last bytes of the file), but other portions (such as the
+/// PageIndex) can be stored elsewhere.
+///
+/// This reader handles reading the footer as well as the non contiguous parts
+/// of the metadata such as the page indexes.
```
Comment (progval) [inline parquet/src/file/metadata/reader.rs:47]: Does it read Bloom Filters too?

Comment (alamb) [reply parquet/src/file/metadata/reader.rs:47]: It does not read Bloom Filters yet. That is a good suggestion. I will file a ticket to consider adding support

Comment (progval) [reply parquet/src/file/metadata/reader.rs:47]:
```suggestion
/// This reader handles reading the footer as well as the non contiguous parts
/// of the metadata such as the page indexes; excluding Bloom Filters.
```

Comment (alamb) [reply parquet/src/file/metadata/reader.rs]: > It does not read Bloom Filters yet. That is a good suggestion. I will file a ticket to consider adding support

Filed https://github.com/apache/arrow-rs/issues/6514 to track the idea. Thank you again @progval

**Comment by alamb** [comment]
Thank you everyone for your suggestions and support.

**Pull request merged**